File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/C/snapman/snapman.c
//============================================================//
//                                                            //
// $ID$                                                       //
//                                                            //
// snapman.c                                                  //
//                                                            //
// Application which uses ONTAPI APIs to get snapshot lists	  //
// and schedules, and take, rename and delete snapshots.      //
//                                                            //
// Copyright 2003 Network Appliance, Inc. All rights		  //
// reserved. Specifications subject to change without notice. // 
//                                                            //
// This SDK sample code is provided AS IS, with no support or //
// warranties of any kind, including but not limited to       //
// warranties of merchantability or fitness of any kind,      //
// expressed or implied.  This code is subject to the license //
// agreement that accompanies the SDK.                        //
//                                                            //
// tab size = 4												  //
//                                                            //
// See Usage() for command-line syntax                        //
//															  //
//============================================================//

//234567890123456789012345678901234567890123456789012345678901234567890123456789

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "../util/netapp.h"
#include "../util/ontapiver.h"

//============================================================//
//
// portability
//
#ifdef WIN32
	#define strcasecmp _stricmp
#endif

//============================================================//

void Usage(char* argv[]);
void GetSchedule(NaServerPtr nsp, const char* volname);
void CreateSnapshot(NaServerPtr nsp, const char* volname, 
									   const char* snapname);
void RenameSnapshot(NaServerPtr nsp, const char* volname,
									   const char* oldsnapname, 
									   const char* newsnapname);
void DeleteSnapshot(NaServerPtr nsp, const char* volname,
									   const char* snapname);
void ListInfo(NaServerPtr nsp, const char* volname);

//============================================================//

int main(int argc, char* argv[])
{
	int			ret = 0;
	char		err[4096];
	char*		filer = argv[2];
	char*		user = argv[3];
	char*		passwd = argv[4];
	NaServer	server;
	
	if (argc < 6) 
		Usage(argv);

	//
	// One-time initialization of system on client
	//
    if (!na_startup(err, sizeof(err))) {
            fprintf(stderr, "Error in na_startup: %s\n", err);
            return -2;
    }

	//
	// request ONTAPI version 1.1
	//
	if (!NaServerInit(&server, filer, user, passwd, 1, 1)) {
		fprintf(stderr, "This filer doesn't support ONTAPI version 1.1\n");
		return -3;
	}

	if (strcasecmp(argv[1], "-g") == 0)
		GetSchedule(&server, argv[5]);
	else if (strcasecmp(argv[1], "-c") == 0)
		CreateSnapshot(&server, argv[5], argv[6]);
	else if (strcasecmp(argv[1], "-r") == 0)
		RenameSnapshot(&server, argv[5], argv[6], argv[7]);
	else if (strcasecmp(argv[1], "-d") == 0)
		DeleteSnapshot(&server, argv[5], argv[6]);
	else if (strcasecmp(argv[1], "-l") == 0)
		ListInfo(&server, argv[5]);
	else
		Usage(argv);	

	na_shutdown();
	return 0;
}

//============================================================//

void Usage(char* argv[])
{
	fprintf(stderr, 
		"Usage: snapman -g <filer> <user> <pw> <vol> \n"
		"               -l <filer> <user> <pw> <vol> \n"
		"               -c <filer> <user> <pw> <vol> <snapshotname> \n"
		"               -r <filer> <user> <pw> <vol> <oldsnapshotname> <newname> \n"
		"               -d <filer> <user> <pw> <vol> <snapshotname> \n\n");
	fprintf(stderr, "E.g. snapman -l filer1 root 6a55w0r9 vol0 \n\n");
	fprintf(stderr, 
		"Use -g to get the snapshot schedule\n"
		"    -l to list snapshot info \n"
		"    -c to create a snapshot \n"
		"    -r to rename one \n"
		"    -d to delete one \n");

	exit(-1);
}

//============================================================//

void GetSchedule(NaServerPtr nsp, const char* volname)
{
	SnapshotSchedule		sssched;

	//
	// get the schedule
	//
	safestrcpy(nsp->volname, volname);

	if (NaServerOpen(nsp) == 0)
		return;

	SnapshotGetSchedule(nsp, &sssched);

	NaServerClose(nsp);

	//
	// print it out
	//
	printf("\n");
	printf("Snapshot schedule for volume %s on filer %s:\n", 
											volname, nsp->filer);
	printf("\n");
	printf("Snapshots are taken on minutes [%s] of each hour (%d kept)\n",
			sssched.whichminutes, sssched.minutes);
	printf("Snapshots are taken on hours [%s] of each day (%d kept)\n", 
			sssched.whichhours, sssched.hours);
	printf("%d nightly snapshots are kept\n", sssched.days);
	printf("%d weekly snapshots are kept\n", sssched.weeks);
	printf("\n");
}

//============================================================//

void CreateSnapshot(NaServerPtr nsp, const char* volname, 
                                        const char* snapname)
{
	strcpy(nsp->volname, volname);

	if (NaServerOpen(nsp) == 0)
		return;

	SnapshotCreate(nsp, snapname);

	NaServerClose(nsp);
}

//============================================================//

void RenameSnapshot(NaServerPtr nsp, const char* volname, 
                                        const char* oldsnapname, 
                                        const char* newsnapname)
{
	strcpy(nsp->volname, volname);

	if (NaServerOpen(nsp) == 0)
		return;

	SnapshotRename(nsp, oldsnapname, newsnapname);

	NaServerClose(nsp);
}

//============================================================//

void DeleteSnapshot(NaServerPtr nsp, const char* volname, 
                                        const char* snapname)
{
	strcpy(nsp->volname, volname);

	if (NaServerOpen(nsp) == 0)
		return;

	SnapshotDelete(nsp, snapname);

	NaServerClose(nsp);
}

//============================================================//


void ListInfo(NaServerPtr nsp, const char* volname)
{
	SnapshotInfoPtr				ssinfo = NULL;
	int							nssitems = 0;

	strcpy(nsp->volname, volname);

	if (NaServerOpen(nsp) == 0)
		return;

	//
	// Get info
	//
	SnapshotListInfo(nsp, &ssinfo, &nssitems);

	//
	// print snapshot info
	//
	if (nssitems) {
		printf("Snapshots on volume %s: \n\n", volname);         
		printf("NAME                    DATE                    BUSY   NBLOCKS CUMNBLOCKS  DEPENDENCY\n");
		printf("-------------------------------------------------------------------------------------\n");
	}
	else {
		printf("No snapshots on volume %s \n\n", volname);
	}
	while (nssitems) {

		char* tmp = asctime(localtime((time_t*)&ssinfo->accesstime));

		tmp[24] = 0;
		printf("%-23s %-24s  %d %10d %10d  %s\n",
				ssinfo->name,
				tmp,
				ssinfo->busy,
				ssinfo->total,
				ssinfo->cumtotal,
				ssinfo->dependency); 
	
		ssinfo++;
		nssitems--;
	}
	printf("\n");
		
	NaServerClose(nsp);
	
}

//============================================================//
//============================================================//
//============================================================//
//============================================================//
//============================================================//

File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/DotNet/CSharp/snapman/snapman.cs
//============================================================//
//                                                            //
// $ID$                                                       //
//                                                            //
// snapman.cs                                                 //
//                                                            //
// Application which uses ONTAPI APIs to get snapshot lists   //
// and schedules, and take, rename and delete snapshots.      //
//                                                            //
// Copyright 2008 NetApp. All rights reserved. Specifications //
// subject to change without notice.                          // 
//                                                            //
// This SDK sample code is provided AS IS, with no support or //
// warranties of any kind, including but not limited to       //
// warranties of merchantability or fitness of any kind,      //
// expressed or implied.  This code is subject to the license //
// agreement that accompanies the SDK.                        //
//                                                            //
//                                                            //
// See Usage() for command-line syntax                        //
//                                                            //
//============================================================//
using System;
using System.Collections.Generic;
using System.Text;
using NetApp.Manage;

namespace snapman
{
    class SnapMan
    {
        static void Usage(string[] args)
        {
           	Console.WriteLine( 
		        "Usage: snapman -g <filer> <user> <pw> <vol> \n" + 
		        "               -l <filer> <user> <pw> <vol> \n" + 
		        "               -c <filer> <user> <pw> <vol> <snapshotname> \n" + 
		        "               -r <filer> <user> <pw> <vol> <oldsnapshotname> <newname> \n" + 
		        "               -d <filer> <user> <pw> <vol> <snapshotname> \n\n");
    	    Console.WriteLine("E.g. snapman -l filer1 root 6a55w0r9 vol0 \n\n");
            Console.WriteLine(
		        "Use -g to get the snapshot schedule\n" + 
		        "    -l to list snapshot info \n" + 
		        "    -c to create a snapshot \n" + 
		        "    -r to rename one \n" + 
		        "    -d to delete one \n");

        	System.Environment.Exit(-1);
        }
        static void CreateSnapshot(String[] args)
        {
            NaServer s;
            string filer = args[1];
            string user = args[2];
            string pwd = args[3];
            string vol = args[4];
            string ssname = args[5];
            NaElement xi, xo;

            try
            {
                s = new NaServer(filer, 1, 0);
                s.Style = NaServer.AUTH_STYLE.LOGIN_PASSWORD;
                s.TransportType = NaServer.TRANSPORT_TYPE.HTTP;
                s.SetAdminUser(user, pwd);

                xi = new NaElement("snapshot-create");
                if (args.Length == 6)
                {
                    xi.AddNewChild("volume", vol);
                    xi.AddNewChild("snapshot", ssname);
                }
                else
                {
                    Console.Error.WriteLine("Invalid number of arguments");
                    Usage(args);
                    System.Environment.Exit(-1);
                }

                xo = s.InvokeElem(xi);
                //
                // print it out
                //
                Console.WriteLine("Snapshot " + ssname + " created for volume " + vol + " on filer " + filer);
            }
            catch (NaException e)
            {
                Console.Error.WriteLine("ERROR: " + e.Message);
            }
        }


        static void DeleteSnapshot(String[] args)
        {
            try
            {
                NaServer s;
                string filer = args[1];
                string user = args[2];
                string pwd = args[3];
                string vol = args[4];
                string ssname = args[5];
                NaElement xi, xo;

                s = new NaServer(filer, 1, 0);
                s.Style = NaServer.AUTH_STYLE.LOGIN_PASSWORD;
                s.TransportType = NaServer.TRANSPORT_TYPE.HTTP;
                s.SetAdminUser(user, pwd);

                xi = new NaElement("snapshot-delete");
                xi.AddNewChild("volume", vol);
                xi.AddNewChild("snapshot", ssname);

                xo = s.InvokeElem(xi);
                //
                // print it out
                //
                Console.WriteLine("Snapshot " + ssname + " deleted from volume " + vol + " on filer " + filer);
            }
            catch (IndexOutOfRangeException e)
            {
                Console.Error.WriteLine("Invalid number of arguments");
                Usage(args);
                Console.Error.WriteLine(e.Message);
                System.Environment.Exit(-1);
            }
            catch (NaException e)
            {
                Console.Error.WriteLine("ERROR: " + e.Message);
            }
        }


        static void RenameSnapshot(String[] args)
        {
            try
            {
                NaServer s;
                string filer = args[1];
                string user = args[2];
                string pwd = args[3];
                string vol = args[4];
                string ssnameOld = args[5];
                string ssnameNew = args[6];
                NaElement xi, xo;

                s = new NaServer(filer, 1, 0);
                s.Style = NaServer.AUTH_STYLE.LOGIN_PASSWORD;
                s.TransportType = NaServer.TRANSPORT_TYPE.HTTP;
                s.SetAdminUser(user, pwd);

                xi = new NaElement("snapshot-rename");
                if (args.Length == 7)
                {
                    xi.AddNewChild("volume", vol);
                    xi.AddNewChild("current-name", ssnameOld);
                    xi.AddNewChild("new-name", ssnameNew);
                }
                else
                {
                    Console.Error.WriteLine("Invalid number of arguments");
                    Usage(args);
                    System.Environment.Exit(-1);
                }

                xo = s.InvokeElem(xi);
                //
                // print it out
                //
                Console.WriteLine("Snapshot " + ssnameOld + " renamed to " + 
                                        ssnameNew + " for volume " + vol + " on filer " + filer);
            }
            catch (IndexOutOfRangeException e)
            {
                Console.Error.WriteLine("ERROR:" + e.Message);
                Usage(args);
                System.Environment.Exit(-1);
            }
            catch (NaException e)
            {
                Console.Error.WriteLine("ERROR: " + e.Message);
            }
        }


        static void ListInfo(String[] args)
        {
            NaServer s;
            string filer = args[1];
            string user = args[2];
            string pwd = args[3];
            string vol = args[4];

            NaElement xi, xo;
            //
            // get the schedule
            //
            try
            {
                s = new NaServer(filer, 1, 0);
                s.Style = NaServer.AUTH_STYLE.LOGIN_PASSWORD;
                s.TransportType = NaServer.TRANSPORT_TYPE.HTTP;
                s.SetAdminUser(user, pwd);

                xi = new NaElement("snapshot-list-info");
                xi.AddNewChild("volume", vol);
                xo = s.InvokeElem(xi);

                System.Collections.IList snapshots = xo.GetChildByName("snapshots").GetChildren();
                System.Collections.IEnumerator snapiter = snapshots.GetEnumerator();
                while (snapiter.MoveNext())
                {
                    NaElement snapshot = (NaElement)snapiter.Current;
                    Console.WriteLine("SNAPSHOT:");
                    int accesstime = snapshot.GetChildIntValue("access-time",0);
                    DateTime datetime = new DateTime(1970,1,1,0,0,0).AddSeconds(accesstime);
                    Console.WriteLine("   NAME \t\t= " + snapshot.GetChildContent("name"));
                    Console.WriteLine("   ACCESS TIME (GMT) \t= " + datetime);
                    Console.WriteLine("   BUSY \t\t= " + snapshot.GetChildContent("busy"));
                    Console.WriteLine("   TOTAL (of 1024B) \t= " + snapshot.GetChildContent("total"));
                    Console.WriteLine("   CUMULATIVE TOTAL (of 1024B) = " + snapshot.GetChildContent("cumulative-total"));
                    Console.WriteLine("   DEPENDENCY \t\t= " + snapshot.GetChildContent("dependency"));
                }
            }
            catch (NaAuthException e)
            {
                Console.Error.WriteLine("Authentication Error : " + e.Message);
            }
            catch (NaApiFailedException e)
            {
                Console.Error.WriteLine("API Failed : " + e.Message );
            }

        }


        static void GetSchedule(String[] args)
        {
            NaServer s;
            string filer = args[1];
            string user = args[2];
            string pwd = args[3];
            string vol = args[4];
            
            NaElement xi, xo;
            //
            // get the schedule
            //
            try
            {
                s = new NaServer(filer, 1, 0);
                s.Style = NaServer.AUTH_STYLE.LOGIN_PASSWORD;
                s.TransportType = NaServer.TRANSPORT_TYPE.HTTP;
                s.SetAdminUser(user, pwd);

                xi = new NaElement("snapshot-get-schedule");
                if (args.Length == 5)
                {
                    xi.AddNewChild("volume", vol);
                }
                else 
                {
                    Console.Error.WriteLine("Invalid number of arguments");
                    Usage(args);
                    System.Environment.Exit(-1);
                }

                xo = s.InvokeElem(xi);
                //
                // print it out
                //
                Console.WriteLine("Snapshot schedule for volume "+ vol +  " on filer " + filer + ":");
                Console.WriteLine("-----------------------------------------------------------------");
                Console.WriteLine("Snapshots are taken on minutes [" + 
                            xo.GetChildIntValue("which-minutes", 0) + "] of each hour (" +
                            xo.GetChildContent("minutes") + " kept)" );
                Console.WriteLine("Snapshots are taken on hours [" +
                            xo.GetChildContent("which-hours") + "] of each day (" +
                            xo.GetChildContent("hours") + " kept)\n");
                Console.WriteLine(xo.GetChildContent("days") + " nightly snapshots are kept\n");
                Console.WriteLine(xo.GetChildContent("weeks") + " weekly snapshots are kept\n");
                Console.WriteLine("\n");
            }
            catch (NaException e)
            {
                Console.WriteLine("ERROR: " + e.Message);
            }
            
        }
        static void Main(string[] args)
        {
            if (args.Length < 5) 
		        Usage(args);

            Console.WriteLine("|--------------------------------------|");
            Console.WriteLine("| Program to Demo use of Snapshot APIs |");
            Console.WriteLine("|--------------------------------------|");
                
            switch (args[0]) {
                case "-g":
                    GetSchedule(args);
                    break;
                case "-c":
                    CreateSnapshot(args);
                    break;
                case "-r":
                    RenameSnapshot(args);
                    break;
                case "-d":
                    DeleteSnapshot(args);
                    break;
                case "-l":
                    ListInfo(args);
                    break;
                default:
                    Usage(args);
                    break;
            }

        }
    }
}
